### DISCUSSION TITLE
Excluding and/or referencing an endpoint on inbound events processed

### DISCUSSION THREAD

**Discussion by SNewtown** [orchetect/swift-midi#189] [Q&A]
I found in the documentation that .addInputConnection can include/specify an endpoint using the toOutputs but can it exclude an endpoint instead? 

And I read in the discussions that you don't recommend using the name (i.e. toOutputs: [.name("device 1")] for an inclusion) - what else could be used to identify an Endpoint if not the name?

Also, is the endpoint referencable in the MIDIEvent (once passed to the listener and messages are being parsed)? In other words, can a CC or SysEx message be identified to an endpoint from the MIDIEvent?

**Comment by orchetect** [comment]
> I found in the documentation that .addInputConnection can include/specify an endpoint using the toOutputs but can it exclude an endpoint instead?

That call has a filter parameter for that exact purpose. [addInputConnection(toOutputs:tag:mode:filter:receiver:)](https://orchetect.github.io/MIDIKit/documentation/midikit/midimanager/addinputconnection(tooutputs:tag:mode:filter:receiver:)-5xxyz). It's generally only useful when using the [.allEndpoints](https://orchetect.github.io/MIDIKit/documentation/midikit/midiconnectionmode/allendpoints) mode on the connection but can be useful to prevent endpoints from being added programmatically too.

> And I read in the discussions that you don't recommend using the name (i.e. toOutputs: [.name("device 1")] for an inclusion) - what else could be used to identify an Endpoint if not the name?

Endpoint unique ID, always if at all possible. [.uniqueID(_:)] (https://orchetect.github.io/MIDIKit/documentation/midikit/midiendpointidentity/uniqueid(_:)) The docs for endpoint criteria are clear on that.
 
> Also, is the endpoint referencable in the MIDIEvent (once passed to the listener and messages are being parsed)? In other words, can a CC or SysEx message be identified to an endpoint from the MIDIEvent?

Use the [`.eventsWithMetadata { event, timestamp, source in }`](https://orchetect.github.io/MIDIKit/documentation/midikit/midireceiver/eventswithmetadata(translatemidi1noteonzerovelocitytonoteoff:_:)) receive handler instead of [`.events { event in }`](https://orchetect.github.io/MIDIKit/documentation/midikit/midireceiver/events(translatemidi1noteonzerovelocitytonoteoff:_:)). Source endpoint will be present when used with an Input Connection, but not with a Virtual Input due to Core MIDI limitations.
